pkg/ipcache/metadata.go

```diff
@@ -79,10 +79,12 @@ func (m *metadata) upsert(prefix string, lbls labels.Labels) {
 	m.Unlock()
 }
 
-func (m *metadata) delete(prefix string) {
-	m.Lock()
-	delete(m.m, prefix)
-	m.Unlock()
+func (ipc *IPCache) RemoveMetadata(prefix string, lbls labels.Labels, src source.Source) {
+	ipc.metadata.Lock()
+	ipc.Lock()
+	_ = ipc.removeLabels(prefix, lbls, src)
```

Comment:
Heads up that this solves cases where there is a clear disassociation of the node first (eg node is deleted from k8s) before associating the node with another resource like a service, fqdn or kube-apiserver.

In order to solve the other order, we need to handle the resulting labels from this function. I think it's reasonable to review this PR as-is since this at least fixes two bugs (deadlock + deletion of the node followed by association with another resource), then we can follow up on the other case.